chalice/deploy/packager.py

```diff
@@ -208,3 +216,373 @@ def _add_chalice_lib_if_needed(self, project_dir, zip_fileobj):
+            if self._valid_lambda_whl(whl.filename):
+                valid_whls.add(whl)
+            else:
+                invalid_whls.add(whl)
+        return valid_whls, invalid_whls
+
+    def _download_dependencies(self, directory, requirements_file):
+        # type: (str, str) -> Tuple[Set[Package], Set[Package]]
+        # Download all dependencies we can, letting pip choose what to download
+        deps = self._download_all_dependencies(requirements_file, directory)
+
+        # Sort the downloaded packages into three categories:
+        # - sdists (Pip could not get a wheel so it gave us a sdist)
+        # - valid whls (lambda compatbile wheel files)
+        # - invalid whls (lambda incompatible wheel files)
+        valid_whls, invalid_whls = self._categorize_whl_files(directory)
+        sdists = deps - valid_whls - invalid_whls
+
+        # Find which packages we do not yet have a valid whl file for. And
+        # try to download them specifically with lambda.
+        missing_whls = sdists | invalid_whls
+        self._download_binary_wheels(missing_whls, directory)
+
+        # Re-count the whl files after the second download pass. Anything
+        # that has a sdist but not a valid whl file is still missing and needs
+        # to be built from source into a wheel file.
+        valid_whls, invalid_whls = self._categorize_whl_files(directory)
+        missing_whls = sdists - valid_whls
+        self._build_sdists(missing_whls, directory)
+
+        # Final pass to find the valid whl files and see if there are any
+        # unmet dependencies left over. At this point there is nothing we can
+        # do about any missing wheel files.
+        valid_whls, _ = self._categorize_whl_files(directory)
+        missing_whls = deps - valid_whls
+        return valid_whls, missing_whls
+
+    def _install_purelib_and_platlib(self, whl, root):
+        # type: (Package, str) -> None
+        # Take a wheel package and the directory it was just unpacked into and
+        # properly unpackage the purelib and platlib subdirectories if they
+        # are present.
+        data_dir = self._osutils.joinpath(root, whl.data_dir)
+        if not self._osutils.directory_exists(data_dir):
+            return
+        unpack_dirs = {'purelib', 'platlib'}
+        data_contents = self._osutils.get_directory_contents(data_dir)
+        for content_name in data_contents:
+            if content_name in unpack_dirs:
+                source = self._osutils.joinpath(data_dir, content_name)
+                self._osutils.copytree(source, root)
+                # No reason to keep the purelib/platlib source directory around
+                # so we delete it to conserve space in the package.
+                self._osutils.rmtree(source)
+
+    def _install_whls(self, src_dir, dst_dir, whls):
+        # type: (str, str, Set[Package]) -> None
+        if os.path.isdir(dst_dir):
+            shutil.rmtree(dst_dir)
+        os.makedirs(dst_dir)
```

Comment:
If you're introducing `OSUtils` as a dependency for this object, the point of that is it abstracts *all* os operations to allow for granular testing, error conditions, etc.  Using `shutil` and `os` directly in the same object defeats that purpose.